File: youtube_auth_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
youtube_auth_setup.py — One-time local OAuth 2.0 authentication helper.

Run this script locally on your machine to authorize your YouTube channel.
It opens a browser window, completes Google OAuth 2.0 authorization,
and outputs your permanent REFRESH TOKEN for use in .env and GitHub Actions Secrets.

Prerequisites:
1. Go to Google Cloud Console (https://console.cloud.google.com/).
2. Enable "YouTube Data API v3".
3. Under "APIs & Services" -> "OAuth consent screen", select "External", add test user (your Google email).
4. Under "Credentials", click "Create Credentials" -> "OAuth client ID" -> "Desktop App".
5. Download JSON as `client_secrets.json` into this project root, OR enter Client ID & Secret when prompted.
"""

import os
import sys
import json

# Cross-platform UTF-8 output
if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")

SCOPES = [
    "https://www.googleapis.com/auth/youtube.upload",
    "https://www.googleapis.com/auth/youtube",
    "https://www.googleapis.com/auth/youtube.force-ssl",
]

CLIENT_SECRETS_FILE = os.path.join(os.path.dirname(__file__), "client_secrets.json")
ENV_FILE = os.path.join(os.path.dirname(__file__), ".env")


def update_env_file(client_id: str, client_secret: str, refresh_token: str):
    """Safely updates or appends YouTube OAuth credentials in the local .env file."""
    env_vars = {}
    if os.path.exists(ENV_FILE):
        with open(ENV_FILE, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    k, v = line.split("=", 1)
                    env_vars[k.strip()] = v.strip()

    env_vars["YOUTUBE_CLIENT_ID"] = client_id
    env_vars["YOUTUBE_CLIENT_SECRET"] = client_secret
    env_vars["YOUTUBE_REFRESH_TOKEN"] = refresh_token

    with open(ENV_FILE, "w", encoding="utf-8") as f:
        for k, v in env_vars.items():
            f.write(f"{k}={v}\n")
    print(f"✅ Updated local .env with YouTube OAuth credentials.")


def main():
    print("=" * 70)
    print("  YouTube Channel OAuth 2.0 Setup (One-Time Token Generator)")
    print("=" * 70)

    from google_auth_oauthlib.flow import InstalledAppFlow

    flow = None
    client_id = None
    client_secret = None

    if os.path.exists(CLIENT_SECRETS_FILE):
        print(f"📁 Found {CLIENT_SECRETS_FILE}! Loading client configuration...")
        with open(CLIENT_SECRETS_FILE, "r", encoding="utf-8") as f:
            secret_data = json.load(f)
            # Support both 'installed' and 'web' client structures
            client_info = secret_data.get("installed") or secret_data.get("web") or {}
            client_id = client_info.get("client_id")
            client_secret = client_info.get("client_secret")

        flow = InstalledAppFlow.from_client_secrets_file(
            CLIENT_SECRETS_FILE,
            scopes=SCOPES,
        )
    else:
        print("\nℹ️  client_secrets.json not found in root folder.")
        client_id = input("Enter your Google OAuth Client ID: ").strip()
        client_secret = input("Enter your Google OAuth Client Secret: ").strip()

        if not client_id or not client_secret:
            print("❌ Client ID and Secret are required to proceed.")
            sys.exit(1)

        client_config = {
            "installed": {
                "client_id": client_id,
                "client_secret": client_secret,
                "auth_uri": "https://accounts.google.com/o/oauth2/auth",
                "token_uri": "https://oauth2.googleapis.com/token",
                "redirect_uris": ["http://localhost:8080/", "http://127.0.0.1:8080/"],
            }
        }
        flow = InstalledAppFlow.from_client_config(
            client_config,
            scopes=SCOPES,
        )

    print("\n🌐 Opening local browser for Google account authorization...")
    print("⚠️  Make sure you log in with the Google account that owns or manages the YouTube channel.")
    print("   If you see a 'Google hasn't verified this app' screen, click 'Advanced' -> 'Go to <app> (unsafe)'.")

    # Run local server to capture the OAuth redirect
    credentials = flow.run_local_server(
        port=8080,
        prompt="consent",
        access_type="offline",
        open_browser=True,
    )

    refresh_token = credentials.refresh_token

    if not refresh_token:
        print("\n⚠️  No refresh token returned! This happens if consent was already granted.")
        print("   To force Google to generate a new refresh token, revoke app access at:")
        print("   https://myaccount.google.com/permissions and run this script again.")
        sys.exit(1)

    print("\n" + "=" * 70)
    print("🎉 OAuth Authorization Successful!")
    print("=" * 70)
    print("\nHere are your persistent credentials:\n")
    print(f"YOUTUBE_CLIENT_ID={client_id}")
    print(f"YOUTUBE_CLIENT_SECRET={client_secret}")
    print(f"YOUTUBE_REFRESH_TOKEN={refresh_token}")

    print("\n" + "-" * 70)
    print("📋 For GitHub Actions Deployment:")
    print("Add these 3 secrets to your GitHub Repository Settings -> Secrets and variables -> Actions:")
    print("  1. YOUTUBE_CLIENT_ID")
    print("  2. YOUTUBE_CLIENT_SECRET")
    print("  3. YOUTUBE_REFRESH_TOKEN")
    print("-" * 70)

    # Save to local .env
    update_env_file(client_id, client_secret, refresh_token)
    print("\nSetup complete! You can now run `py youtube_uploader.py` or test live uploads.")


if __name__ == "__main__":
    main()

File: youtube_auth_setup.js
/**
 * youtube_auth_setup.js — Standalone Node.js OAuth 2.0 helper for YouTube API
 */
const http = require('http');
const { exec } = require('child_process');
const fs = require('fs');
const path = require('path');
const readline = require('readline');

const SCOPES = [
  'https://www.googleapis.com/auth/youtube.upload',
  'https://www.googleapis.com/auth/youtube',
  'https://www.googleapis.com/auth/youtube.force-ssl'
];

async function ask(query) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  return new Promise((resolve) => rl.question(query, (ans) => {
    rl.close();
    resolve(ans.trim());
  }));
}

async function main() {
  console.log('='.repeat(70));
  console.log('   YouTube Channel OAuth 2.0 Setup (One-Time Token Generator)');
  console.log('='.repeat(70));

  let clientId = process.argv[2] || process.env.YOUTUBE_CLIENT_ID;
  let clientSecret = process.argv[3] || process.env.YOUTUBE_CLIENT_SECRET;

  const clientSecretsPath = path.join(__dirname, 'client_secrets.json');
  if (fs.existsSync(clientSecretsPath)) {
    try {
      const data = JSON.parse(fs.readFileSync(clientSecretsPath, 'utf8'));
      const info = data.installed || data.web || {};
      clientId = clientId || info.client_id;
      clientSecret = clientSecret || info.client_secret;
    } catch (e) {}
  }

  if (!clientId) {
    clientId = await ask('\nEnter your Google OAuth Client ID: ');
  }
  if (!clientSecret) {
    clientSecret = await ask('Enter your Google OAuth Client Secret: ');
  }

  if (!clientId || !clientSecret) {
    console.error('❌ Client ID and Secret are required.');
    process.exit(1);
  }

  const port = 8080;
  const redirectUri = `http://localhost:${port}/`;

  // Direct code exchange fallback
  const codeArgIdx = process.argv.findIndex(a => a === '--code');
  let manualCode = codeArgIdx !== -1 ? process.argv[codeArgIdx + 1] : null;
  if (manualCode) {
    if (manualCode.includes('code=')) {
      try {
        const u = new URL(manualCode);
        manualCode = u.searchParams.get('code');
      } catch (e) {
        const match = manualCode.match(/code=([^&]+)/);
        if (match) manualCode = decodeURIComponent(match[1]);
      }
    }
    console.log('\n⏳ Exchanging provided authorization code for tokens...');
    const tokenRes = await fetch('https://oauth2.googleapis.com/token', {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({
        code: manualCode,
        client_id: clientId,
        client_secret: clientSecret,
        redirect_uri: redirectUri,
        grant_type: 'authorization_code'
      }).toString()
    });
    const tokenData = await tokenRes.json();
    if (!tokenRes.ok || tokenData.error) {
      console.error('❌ Failed to obtain tokens:', tokenData);
      process.exit(1);
    }
    const refreshToken = tokenData.refresh_token;
    if (!refreshToken) {
      console.warn('⚠️ No refresh token was returned.');
      process.exit(1);
    }
    console.log('🎉 OAuth Authorization Complete!');
    console.log(`\nYOUTUBE_REFRESH_TOKEN=${refreshToken}`);
    // Sync to .env and gh
    const envPath = path.join(__dirname, '.env');
    let envContent = fs.existsSync(envPath) ? fs.readFileSync(envPath, 'utf8') : '';
    const setEnvVar = (content, key, val) => {
      const regex = new RegExp(`^${key}=.*$`, 'm');
      if (regex.test(content)) return content.replace(regex, `${key}=${val}`);
      return content + (content.endsWith('\n') || !content ? '' : '\n') + `${key}=${val}\n`;
    };
    envContent = setEnvVar(envContent, 'YOUTUBE_CLIENT_ID', clientId);
    envContent = setEnvVar(envContent, 'YOUTUBE_CLIENT_SECRET', clientSecret);
    envContent = setEnvVar(envContent, 'YOUTUBE_REFRESH_TOKEN', refreshToken);
    fs.writeFileSync(envPath, envContent, 'utf8');
    const ghBin = fs.existsSync('C:\\Program Files\\GitHub CLI\\gh.exe')
      ? '"C:\\Program Files\\GitHub CLI\\gh.exe"'
      : 'gh';
    exec(`${ghBin} secret set YOUTUBE_CLIENT_ID --body "${clientId}"`, () => {
      exec(`${ghBin} secret set YOUTUBE_CLIENT_SECRET --body "${clientSecret}"`, () => {
        exec(`${ghBin} secret set YOUTUBE_REFRESH_TOKEN --body "${refreshToken}"`, () => {
          console.log('✅ Updated all GitHub Secrets!');
          process.exit(0);
        });
      });
    });
    return;
  }

  const authUrl = `https://accounts.google.com/o/oauth2/v2/auth?client_id=${encodeURIComponent(
    clientId
  )}&redirect_uri=${encodeURIComponent(redirectUri)}&response_type=code&scope=${encodeURIComponent(
    SCOPES.join(' ')
  )}&access_type=offline&prompt=consent`;

  const server = http.createServer(async (req, res) => {
    try {
      const reqUrl = new URL(req.url, `http://localhost:${port}`);
      const code = reqUrl.searchParams.get('code');
      const error = reqUrl.searchParams.get('error');

      if (error) {
        res.writeHead(400, { 'Content-Type': 'text/html' });
        res.end(`<h2>Authorization Error: ${error}</h2>`);
        console.error(`\n❌ Authorization failed: ${error}`);
        server.close();
        process.exit(1);
      }

      if (!code) {
        res.writeHead(404);
        res.end();
        return;
      }

      res.writeHead(200, { 'Content-Type': 'text/html' });
      res.end(`
        <div style="font-family:sans-serif; text-align:center; padding:50px;">
          <h1 style="color:#16a34a;">🎉 Authorization Successful!</h1>
          <p>You can close this tab and return to the application.</p>
        </div>
      `);

      server.close();

      console.log('\n⏳ Exchanging authorization code for tokens...');
      const tokenRes = await fetch('https://oauth2.googleapis.com/token', {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: new URLSearchParams({
          code: code,
          client_id: clientId,
          client_secret: clientSecret,
          redirect_uri: redirectUri,
          grant_type: 'authorization_code'
        }).toString()
      });

      const tokenData = await tokenRes.json();
      if (!tokenRes.ok || tokenData.error) {
        console.error('❌ Failed to obtain tokens:', tokenData);
        process.exit(1);
      }

      const refreshToken = tokenData.refresh_token;
      if (!refreshToken) {
        console.warn('⚠️ No refresh token was returned.');
        console.warn('Go to https://myaccount.google.com/permissions to revoke access, then run again.');
        process.exit(1);
      }

      console.log('\n' + '='.repeat(70));
      console.log('🎉 OAuth Authorization Complete!');
      console.log('='.repeat(70));
      console.log(`\nYOUTUBE_CLIENT_ID=${clientId}`);
      console.log(`YOUTUBE_CLIENT_SECRET=${clientSecret}`);
      console.log(`YOUTUBE_REFRESH_TOKEN=${refreshToken}`);

      // Save to local .env
      const envPath = path.join(__dirname, '.env');
      let envContent = '';
      if (fs.existsSync(envPath)) {
        envContent = fs.readFileSync(envPath, 'utf8');
      }
      const setEnvVar = (content, key, val) => {
        const regex = new RegExp(`^${key}=.*$`, 'm');
        if (regex.test(content)) {
          return content.replace(regex, `${key}=${val}`);
        }
        return content + (content.endsWith('\n') || !content ? '' : '\n') + `${key}=${val}\n`;
      };
      envContent = setEnvVar(envContent, 'YOUTUBE_CLIENT_ID', clientId);
      envContent = setEnvVar(envContent, 'YOUTUBE_CLIENT_SECRET', clientSecret);
      envContent = setEnvVar(envContent, 'YOUTUBE_REFRESH_TOKEN', refreshToken);
      fs.writeFileSync(envPath, envContent, 'utf8');
      console.log('\n✅ Updated local .env file.');

      // Update GitHub Secrets via GitHub CLI
      console.log('🔄 Syncing secrets to GitHub repository via GitHub CLI...');
      const ghBin = fs.existsSync('C:\\Program Files\\GitHub CLI\\gh.exe')
        ? '"C:\\Program Files\\GitHub CLI\\gh.exe"'
        : 'gh';
      try {
        exec(`${ghBin} secret set YOUTUBE_CLIENT_ID --body "${clientId}"`, () => {
          exec(`${ghBin} secret set YOUTUBE_CLIENT_SECRET --body "${clientSecret}"`, () => {
            exec(`${ghBin} secret set YOUTUBE_REFRESH_TOKEN --body "${refreshToken}"`, (err) => {
              if (!err) {
                console.log('✅ Successfully updated all 3 GitHub Secrets (YOUTUBE_CLIENT_ID, YOUTUBE_CLIENT_SECRET, YOUTUBE_REFRESH_TOKEN)!');
              } else {
                console.log('ℹ️ Updated local .env. GitHub CLI secret sync error:', err.message);
              }
              process.exit(0);
            });
          });
        });
      } catch (err) {
        console.error('GitHub secret sync error:', err);
        process.exit(0);
      }

    } catch (err) {
      console.error('Error handling redirect:', err);
      server.close();
      process.exit(1);
    }
  });

  server.listen(port, () => {
    console.log(`\n🌐 Opening browser for authorization...\n`);
    console.log(`If it doesn't open automatically, open this URL:`);
    console.log(`👉 ${authUrl}\n`);
    exec(`start "" "${authUrl}"`);
  });
}

main().catch(console.error);
